provider/pdns.go

```diff
@@ -0,0 +1,369 @@
+			}
+			return endpoints[i].DNSName < endpoints[j].DNSName
+		})
+
+	zones, _, err := p.client.ListZones()
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort the zone by length of the name in descending order, we use this
+	// property later to ensure we add a record to the longest matching zone
+
+	sort.SliceStable(zones, func(i, j int) bool { return len(zones[i].Name) > len(zones[j].Name) })
+
+	// NOTE: Complexity of this loop is O(Zones*Endpoints).
+	// A possibly faster implementation would be a search of the reversed
+	// DNSName in a trie of Zone names, which should be O(Endpoints), but at this point it's not
+	// necessary.
+	for _, zone := range zones {
+		zone.Rrsets = []pgo.RrSet{}
+		for i := 0; i < len(endpoints); {
+			ep := endpoints[0]
+			dnsname := ensureTrailingDot(ep.DNSName)
+			if strings.HasSuffix(dnsname, zone.Name) {
+				// The assumption here is that there will only ever be one target
+				// per (ep.DNSName, ep.RecordType) tuple, which holds true for
+				// external-dns v5.0.0-alpha onwards
+				records := []pgo.Record{}
+				for _, t := range ep.Targets {
+					records = append(records, pgo.Record{Content: t})
+				}
+				rrset := pgo.RrSet{
+					Name:       dnsname,
+					Type_:      ep.RecordType,
+					Records:    records,
+					Changetype: string(changetype),
+				}
+
+				// DELETEs explicitly forbid a TTL, therefore only PATCHes need the TTL
+				if changetype == PdnsReplace {
+					if int64(ep.RecordTTL) > int64(math.MaxInt32) {
+						return nil, errors.New("Value of record TTL overflows, limited to int32")
+					}
+					if ep.RecordTTL == 0 {
+						// No TTL was sepecified for the record, we use the default
+						rrset.Ttl = int32(defaultTTL)
+					} else {
+						rrset.Ttl = int32(ep.RecordTTL)
+					}
+				}
+
+				zone.Rrsets = append(zone.Rrsets, rrset)
+
+				// "pop" endpoint if it's matched
+				endpoints = append(endpoints[0:i], endpoints[i+1:]...)
+			} else {
+				// If we didn't pop anything, we move to the next item in the list
+				i++
+			}
+
+		}
+
+		if len(zone.Rrsets) > 0 {
+			zonelist = append(zonelist, zone)
+		}
+
+	}
+
+	// If we still have some endpoints left, it means we couldn't find a matching zone for them
+	// We warn instead of hard fail here because we don't want a misconfig to cause everything to go down
+	if len(endpoints) > 0 {
+		log.Warnf("No matching zones were found for the following endpoints: %+v", endpoints)
+	}
+
+	log.Debugf("Zone List generated from Endpoints: %+v", zonelist)
+
+	return zonelist, nil
+}
+
+// mutateRecords takes a list of endpoints and creates, replaces or deletes them based on the changetype
+func (p *PDNSProvider) mutateRecords(endpoints []*endpoint.Endpoint, changetype pdnsChangeType) error {
+	zonelist, err := p.ConvertEndpointsToZones(endpoints, changetype)
+	if err != nil {
+		return err
+	}
+	for _, zone := range zonelist {
+		jso, err := json.Marshal(zone)
+		if err != nil {
+			log.Debugf("JSON Marshal for zone struct failed!")
```

Comment:
I think we should return the error if we fail to unmarshal. It's not so much of a common problem but I think that just logging is pretty dangerous. Also, the log is at debug level which means that it will pass completely unnoticed in case with a lower debug level.